source/common/router/config_impl.cc

```diff
@@ -71,20 +71,30 @@ ShadowPolicyImpl::ShadowPolicyImpl(const envoy::api::v2::RouteAction& config) {
 
 HashPolicyImpl::HashPolicyImpl(
     const Protobuf::RepeatedPtrField<envoy::api::v2::RouteAction::HashPolicy>& hash_policy)
-    : header_name_(hash_policy[0].header().header_name()) {
+    : header_name_(hash_policy[0].header().header_name()),
+      hash_ip_(hash_policy[0].connection_properties().source_ip()) {
   // TODO(htuch): Add support for multiple hash policies, #1422.
   ASSERT(hash_policy.size() == 1);
-  // TODO(htuch): Add support for cookie and source IP hash policies, #1295 and
-  // #1296.
+  // TODO(htuch): Add support for cookie hash policies, #1295
   ASSERT(hash_policy[0].policy_specifier_case() ==
-         envoy::api::v2::RouteAction::HashPolicy::kHeader);
+             envoy::api::v2::RouteAction::HashPolicy::kHeader ||
+         hash_policy[0].policy_specifier_case() ==
+             envoy::api::v2::RouteAction::HashPolicy::kConnectionProperties);
 }
 
-Optional<uint64_t> HashPolicyImpl::generateHash(const Http::HeaderMap& headers) const {
+Optional<uint64_t> HashPolicyImpl::generateHash(const Network::Address::Instance* downstream_addr,
+                                                const Http::HeaderMap& headers) const {
   Optional<uint64_t> hash;
-  const Http::HeaderEntry* header = headers.get(header_name_);
-  if (header) {
-    hash.value(HashUtil::xxHash64(header->value().c_str()));
+  if (!header_name_.get().empty()) {
+    const Http::HeaderEntry* header = headers.get(header_name_);
+    if (header) {
+      hash.value(HashUtil::xxHash64(header->value().c_str()));
+    }
+  } else if (hash_ip_ && downstream_addr) {
```

Comment:
What about the idea of these being cumulative? I.e. I can make the hash a function of both header and downstream IP rather than either/or? I think this was the idea of making the policy a `repeated`.

Reply:
Yeah this is the one part of this change that I find a little strange. 
- Cumulative also makes sense to me.
- It seems odd to trigger header matching based on the empty string. When we translate the repeated is there  a way to translate in such a way that we don't need to check on empty string and perhaps just translate to  generic interface with an evaluate() function like we do in a few other places? (Not sure of details here but just a thought).

Reply:
That's #1422, though it wouldn't be too bad to implement as part of this PR.